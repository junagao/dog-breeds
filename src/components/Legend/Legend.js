import React from 'react';
import PropTypes from 'prop-types';
import { v4 as uuidv4 } from 'uuid';
import {
  TableContainer,
  Table,
  TableThead,
  TableTh,
  TableTr,
  TableTd,
} from './Legend.styles';

const Legend = ({ data }) => {
  const renderLegend = data.map(({ label, value, color, count }) => {
    return (
      <TableTr key={uuidv4()}>
        <TableTd style={{ backgroundColor: color }} />
        <TableTd>{label}</TableTd>
        <TableTd>{(Math.round(value * 100) / 100).toFixed(2)}%</TableTd>
        <TableTd>{count}</TableTd>
      </TableTr>
    );
  });
  return (
    <TableContainer>
      <Table data-testid="legend-table">
        <TableThead>
          <TableTr>
            <TableTh>Color</TableTh>
            <TableTh>Breed</TableTh>
            <TableTh>Percentage</TableTh>
            <TableTh>Image Count</TableTh>
          </TableTr>
        </TableThead>
        <tbody>{renderLegend}</tbody>
      </Table>
    </TableContainer>
  );
};

Legend.propTypes = {
  data: PropTypes.arrayOf(
    PropTypes.shape({
      label: PropTypes.string,
      value: PropTypes.number,
      color: PropTypes.string,
      count: PropTypes.number,
    }),
  ).isRequired,
};

export default Legend;
