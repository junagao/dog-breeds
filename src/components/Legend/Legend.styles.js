import styled from 'styled-components';

export const TableContainer = styled.div`
  max-width: 46rem;
  padding-top: 3rem;
  margin-bottom: 1rem;
  margin: 0 auto;
  @media (min-width: 1280px) {
    max-width: 28rem;
  }
`;

export const Table = styled.table`
  font-size: 0.85rem;
  text-align: left;
  border-radius: 0.25rem 0.25rem 0 0;
  border-collapse: separate;
  border-spacing: 8px;
  table-layout: fixed;
`;

export const TableThead = styled.thead`
  display: table-header-group;
  vertical-align: middle;
  width: 50%;
`;

export const TableTh = styled.th`
  color: rgba(0, 0, 0, 0.85);
  display: table-cell;
  font-weight: 500;
  font-size: 1rem;
  text-align: left;
  transition: background 0.3s ease;
  padding: 0.25rem;
  width: 25%;
  &:first-child {
    width: 3%;
  }
`;

export const TableTr = styled.tr`
  display: table-row;
  position: relative;
  vertical-align: middle;
  transition: background 0.3s ease;
`;

export const TableTd = styled.td`
  padding: 0.5rem;
  color: rgba(0, 0, 0, 0.65);
  &:first-child {
    border-radius: 0.25rem;
    width: 3%;
  }
`;
