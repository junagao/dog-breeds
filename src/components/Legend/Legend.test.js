import React from 'react';
import { render } from '@testing-library/react';
import Legend from './Legend';

describe('Legend', () => {
  const mockDogs = [
    {
      breed: 'chihuahua',
      images: [
        'https://images.dog.ceo/breeds/chihuahua/n02085620_10074.jpg',
        'https://images.dog.ceo/breeds/chihuahua/n02085620_10131.jpg',
        'https://images.dog.ceo/breeds/chihuahua/n02085620_10621.jpg',
        'https://images.dog.ceo/breeds/chihuahua/n02085620_1073.jpg',
        'https://images.dog.ceo/breeds/chihuahua/n02085620_10976.jpg',
        'https://images.dog.ceo/breeds/chihuahua/n02085620_11140.jpg',
      ],
      dogCount: 6,
    },
    {
      breed: 'collie',
      images: [
        'https://images.dog.ceo/breeds/collie-border/n02106166_1031.jpg',
        'https://images.dog.ceo/breeds/collie-border/n02106166_1032.jpg',
        'https://images.dog.ceo/breeds/collie-border/n02106166_1055.jpg',
        'https://images.dog.ceo/breeds/collie-border/n02106166_1056.jpg',
        'https://images.dog.ceo/breeds/collie-border/n02106166_1059.jpg',
        'https://images.dog.ceo/breeds/collie-border/n02106166_1128.jpg',
        'https://images.dog.ceo/breeds/collie-border/n02106166_1133.jpg',
      ],
      dogCount: 7,
    },
    {
      breed: 'husky',
      images: [
        'https://images.dog.ceo/breeds/husky/20180901_150234.jpg',
        'https://images.dog.ceo/breeds/husky/20180904_185604.jpg',
        'https://images.dog.ceo/breeds/husky/20180924_193829.jpg',
      ],
      dogCount: 3,
    },
  ];
  test('matches snapshot', () => {
    const { container } = render(<Legend data={mockDogs} />);
    expect(container).toMatchSnapshot();
  });
  test('displays a table with color, breed, Image Count and Percentage as headers', () => {
    const { getByTestId, getByText } = render(<Legend data={mockDogs} />);
    expect(getByTestId('legend-table')).toBeInTheDocument();
    expect(getByText('Color')).toBeInTheDocument();
    expect(getByText('Breed')).toBeInTheDocument();
    expect(getByText('Image Count')).toBeInTheDocument();
    expect(getByText('Percentage')).toBeInTheDocument();
  });
});
