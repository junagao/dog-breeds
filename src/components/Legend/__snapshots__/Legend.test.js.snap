// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Legend matches snapshot 1`] = `
.c0 {
  max-width: 46rem;
  padding-top: 3rem;
  margin-bottom: 1rem;
  margin: 0 auto;
}

.c1 {
  font-size: 0.85rem;
  text-align: left;
  border-radius: 0.25rem 0.25rem 0 0;
  border-collapse: separate;
  border-spacing: 8px;
  table-layout: fixed;
}

.c2 {
  display: table-header-group;
  vertical-align: middle;
  width: 50%;
}

.c4 {
  color: rgba(0,0,0,0.85);
  display: table-cell;
  font-weight: 500;
  font-size: 1rem;
  text-align: left;
  -webkit-transition: background 0.3s ease;
  transition: background 0.3s ease;
  padding: 0.25rem;
  width: 25%;
}

.c4:first-child {
  width: 3%;
}

.c3 {
  display: table-row;
  position: relative;
  vertical-align: middle;
  -webkit-transition: background 0.3s ease;
  transition: background 0.3s ease;
}

.c5 {
  padding: 0.5rem;
  color: rgba(0,0,0,0.65);
}

.c5:first-child {
  border-radius: 0.25rem;
  width: 3%;
}

@media (min-width:1280px) {
  .c0 {
    max-width: 28rem;
  }
}

<div>
  <div
    class="c0"
  >
    <table
      class="c1"
      data-testid="legend-table"
    >
      <thead
        class="c2"
      >
        <tr
          class="c3"
        >
          <th
            class="c4"
          >
            Color
          </th>
          <th
            class="c4"
          >
            Breed
          </th>
          <th
            class="c4"
          >
            Percentage
          </th>
          <th
            class="c4"
          >
            Image Count
          </th>
        </tr>
      </thead>
      <tbody>
        <tr
          class="c3"
        >
          <td
            class="c5"
          />
          <td
            class="c5"
          />
          <td
            class="c5"
          >
            NaN
            %
          </td>
          <td
            class="c5"
          />
        </tr>
        <tr
          class="c3"
        >
          <td
            class="c5"
          />
          <td
            class="c5"
          />
          <td
            class="c5"
          >
            NaN
            %
          </td>
          <td
            class="c5"
          />
        </tr>
        <tr
          class="c3"
        >
          <td
            class="c5"
          />
          <td
            class="c5"
          />
          <td
            class="c5"
          >
            NaN
            %
          </td>
          <td
            class="c5"
          />
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;
